src/alloc.rs

```diff
@@ -94,74 +82,113 @@ unsafe fn dealloc(ptr: *mut u8, layout: Layout) {
     let headerptr = (ptr as *mut BlockHeader).sub(1);
     let header = ::std::ptr::read(headerptr);
 
-    if !header.prev.is_null() {
-        let mut prev_block = ::std::ptr::read(header.prev);
-        prev_block.next = header.next;
-        ::std::ptr::write(header.prev, prev_block);
-    }
-
-    if !header.next.is_null() {
-        let mut next_block = ::std::ptr::read(header.next);
-        next_block.prev = header.prev;
-        ::std::ptr::write(header.next, next_block);
-    }
+    BlockHeader::patch_next(header.prev, header.next);
+    BlockHeader::patch_prev(header.next, header.prev);
 
     if headerptr == LAST_ALLOCED {
         LAST_ALLOCED = header.prev
     }
-    let baseptr = ptr.sub(uoff);
-    System.dealloc(baseptr, l2);
+    System.dealloc(ptr.sub(uoff), l2);
     ALLOC_LOCK.unlock();
 }
 
 #[inline]
 unsafe fn realloc(ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
     ALLOC_LOCK.lock();
-    let (l2, uoff) = Layout::new::<BlockHeader>().extend(layout).unwrap();
     let old_hp = (ptr as *mut BlockHeader).sub(1);
-    let old_header = ::std::ptr::read(old_hp);
-    let newptr = System.realloc(ptr.sub(uoff), l2, new_size + uoff).add(uoff);
-
-    let new_hp = (newptr as *mut BlockHeader).sub(1);
-    let header = BlockHeader {
-        prev: old_header.prev,
-        next: old_header.next,
-        size: NonZeroUsize::new_unchecked(new_size),
-        type_: old_header.type_,
-    };
-
-    if !header.prev.is_null() {
-        let mut prev_block = ::std::ptr::read(header.prev);
-        prev_block.next = new_hp;
-        ::std::ptr::write(header.prev, prev_block);
-    }
+    let header = ::std::ptr::read(old_hp);
 
-    if !header.next.is_null() {
-        let mut next_block = ::std::ptr::read(header.next);
-        next_block.prev = new_hp;
-        ::std::ptr::write(header.next, next_block);
-    }
+    let (l2, uoff) = Layout::new::<BlockHeader>().extend(layout).unwrap();
+    let ptr = System.realloc(ptr.sub(uoff), l2, new_size + uoff).add(uoff);
 
+    let new_hp = (ptr as *mut BlockHeader).sub(1);
+    header.metadata().size = (new_size as u64).into();
+
+    BlockHeader::patch_next(header.prev, new_hp);
+    BlockHeader::patch_prev(header.next, new_hp);
     ::std::ptr::write(new_hp, header);
 
     if old_hp == LAST_ALLOCED {
         LAST_ALLOCED = new_hp
     }
     ALLOC_LOCK.unlock();
-    newptr
+
+    ptr
 }
 
-pub(crate) struct BlockHeader {
+pub struct BlockHeader {
     prev: *mut BlockHeader,
     next: *mut BlockHeader,
-    size: NonZeroUsize,
-    pub(crate) type_: BlockType,
+    metadata: [u8; SIZE_BLOCKMETADATA],
+}
+
+const SIZE_BLOCKMETADATA: usize = 16;
+
+#[derive(PackedStruct, Default, Debug)]
+#[packed_struct(bit_numbering = "msb0")]
+pub struct BlockMetadata {
+    /// The size of the block (excluding its header)
+    #[packed_field(size_bits = "62", endian = "msb")]
```

Comment:
There's an implicit assumption here that the world is a 64-bit integer. I personally like to add some sort of guard to every such assumption so that if/when we move to different word sizes every part of the program that needs to be reconsidered is explicitly flagged.

Reply:
I agree with this in principle. There already exists a global check on the word size inside the library. Are you suggesting to duplicate this check to places where we make implicit assumptions that a word is 64-bit?

Reply:
Yep.

Reply:
Fixed: 1bd2ee26a71cc4e1aaa443eeaba44d4ee7c90cf4